In [3]:
import sys
import numpy as np
import matplotlib 

print("python:", sys.version)
print("Numpy:", np._version)
print("Matplotlib: ", matplotlib.__version__)

python: 3.9.13 (main, Aug 25 2022, 23:51:50) [MSC v.1916 64 bit (AMD64)]
Numpy: <module 'numpy._version' from 'C:\\Users\\Garnet\\anaconda3\\lib\\site-packages\\numpy\\_version.py'>
Matplotlib:  3.5.2


Densley connected, feed forward, multilayered perceptron model

In [6]:
#inputs from 3 neurons in the previous layer 
#numbers made up for neurons 
inputs = [1, 2, 3]
weights = [0.2, 0.8, -0.5]
# there are 3 connections to the neuron but only one bias associated with the neuron after the input layer 
# the output layer has 4 inputs in this example 
bias  = 2
output = inputs[0]*weights[0] + inputs[1]*weights[1] + inputs[2]*weights[2] + bias
print(output)

2.3


In [8]:
#inputs from 3 neurons in the previous layer 
#numbers made up for neurons
# the output layer has 4 inputs in this example 
inputs = [1, 2, 3, 2.5]
weights = [0.2, 0.8, -0.5, 1.0] 
bias  = 2
output = inputs[0]*weights[0] + inputs[1]*weights[1] + inputs[2]*weights[2] + inputs[3]*weights[3] + bias
print(output)

4.8


In [9]:
#modelling 3 neurons with 4 of the same inputs  
#there is going to be 3 unique weight sets and each weight set is going to have 4 values/ weights
#3 unique biases  
#numbers made up for neurons
# the output layer has 4 inputs in this example 
#changing the output values 
#input data will be data from the chosen features from the basketball.csv dataset
#changing and tuning the inputs will be done through changing and adjusting weights and biases to get the desired output
inputs = [1, 2, 3, 2.5]#vector


weights = [[0.2, 0.8, -0.5, 1.0],
          [0.5, -0.91, 0.26, -0.5],
          [-0.26, -0.27, 0.17, 0.87]]#matrix of vectors

biases = [2, 3, 0.5]#vector

...
layer_outputs = [] #Output of current layer
#zip combines two lists together element-wise
#the 0th element of weights and biases and then iterating over all the other lists
for neuron_weights, neuron_bias, in zip(weights, biases):
    neuron_output = 0 # Output of given neuron
    for n_input, weight in zip(inputs, neuron_weights):
        neuron_output += n_input*weight
    neuron_output+= neuron_bias
    layer_outputs.append(neuron_output)
    
print(layer_outputs)
...


[4.8, 1.21, 2.385]


Adjusting weights and biases to get the desired outputs is important 

Activation functions determine that final output before it becomes an input to another layer or the output of the neural network
weight changes the magnitude of the output 
bias offsets the output node
y=mx+b for a line

Array l = [1,5,6,2] #list numbpi
shape:(4,)
Type: 1D array, Vector 

Array: lol = [[1,5,6,2], [3,2,1,3]] (list of lists = lol)Shape:(2 (amount of lists),4 (number of elements in each list)) Type: 2D Array, Matrix Arrays have to be homolegous which means each element of the list has to be the same dimension a matrix is a rectangular array 

Array: lolol = [[[1,5,6,2],[3,2,1,3]],[[5,2,1,2],[6,4,8,4]],[[2,8,5,3],[1,1,9,4]]]
    shape : (3, 2, 4)
        type: 3D Array

A tensor is an object that can be represented as an array in deep learning . A tensor is not an array

a = [1,2,3]
b = [2,3,4]

dot_product = a[0]*b[0] + a[1]*b[1] + a[2]*b[2]

a.b = [1,2,3].[2,3,4] = 1.2 + 2.3 + 3.4 = 20

The dot product of two vectors gives a scalar single value